File: src/com/alchitry/labs/gui/UndoRedo.java
package com.alchitry.labs.gui;

import org.eclipse.swt.events.VerifyEvent;
import org.eclipse.swt.events.VerifyListener;

import java.util.Stack;

public class UndoRedo implements VerifyListener {
	private StyledCodeEditor editor;
	private Stack<Edit> undoStack;
	private Stack<Edit> redoStack;
	private boolean editing;
	private boolean skip;
	private long lastEdit = 0;

	private class Edit {
		public int position;
		public String text;

		public Edit() {
			position = editor.getCaretOffset();
			text = editor.getText();
		}
	}

	public boolean isEditing() {
		return editing;
	}

	public UndoRedo(StyledCodeEditor editor) {
		this.editor = editor;
		undoStack = new Stack<>();
		redoStack = new Stack<>();
		editing = false;
		skip = false;
	}

	public void skipNext() {
		skip = true;
	}

	@Override
	public void verifyText(VerifyEvent e) {
		if (!skip && !editing) {
			long now = System.currentTimeMillis();
			if ((lastEdit + 1000) < now) {
				lastEdit = now;
				undoStack.push(new Edit());
				redoStack.clear();
			}
		} 
		editing = skip = false;
	}

	private void replace(Stack<Edit> popStack, Stack<Edit> pushStack) {
		if (!editing && popStack.size() > 0) {
			Edit edit = popStack.pop();

			pushStack.push(new Edit());

			editing = true;
			editor.replaceTextRange(0, editor.getCharCount(), edit.text);
			editor.setCaretOffset(edit.position);
			editor.update();
		}
	}

	public void undo() {
		replace(undoStack, redoStack);
	}

	public void redo() {
		replace(redoStack, undoStack);
	}

	public boolean canUndo() {
		return !undoStack.empty();
	}

	public boolean canRedo() {
		return !redoStack.empty();
	}

}

File: src/com/alchitry/labs/gui/util/UndoRedo.kt
package com.alchitry.labs.gui.util

import com.alchitry.labs.gui.StyledCodeEditor
import org.eclipse.swt.events.VerifyEvent
import org.eclipse.swt.events.VerifyListener
import java.util.*

class UndoRedo(private val editor: StyledCodeEditor) : VerifyListener {
    private val undoStack: Stack<Edit> = Stack()
    private val redoStack: Stack<Edit> = Stack()
    var isEditing: Boolean = false
        private set
    private var skip: Boolean = false
    private var lastEdit: Long = 0

    private inner class Edit {
        var position: Int = editor.caretOffset
        var text: String = editor.text
    }

    fun skipNext() {
        skip = true
    }

    override fun verifyText(e: VerifyEvent) {
        if (!skip && !isEditing) {
            val now = System.currentTimeMillis()
            if (lastEdit + 1000 < now) {
                lastEdit = now
                undoStack.push(Edit())
                redoStack.clear()
            }
        }
        skip = false
        isEditing = skip
    }

    private fun replace(popStack: Stack<Edit>, pushStack: Stack<Edit>) {
        if (!isEditing && popStack.size > 0) {
            val edit = popStack.pop()
            pushStack.push(Edit())
            isEditing = true
            editor.text = edit.text
            editor.caretOffset = edit.position
            editor.update()
        }
    }

    fun undo() {
        replace(undoStack, redoStack)
    }

    fun redo() {
        replace(redoStack, undoStack)
    }

    fun canUndo(): Boolean {
        return !undoStack.empty()
    }

    fun canRedo(): Boolean {
        return !redoStack.empty()
    }
}